feat: add thirdparty api specification and supporting docs

lewisdaly opened the pull request:
Closes #70 

This PR adds the Third Party API specification, along with supporting documentation, sequences etc.

TODOS:
- [x] Update the OpenAPI files to the latest version
- [x] Convert Michael's document to a markdown version
- [x] update data model for `POST /thirdpartyRequests/authorizations` based on @ehenrka feedback: https://github.com/mojaloop/mojaloop-specification/issues/70#issuecomment-865013600

elnyry-sam-k commented:
Thanks for this @lewisdaly - having the readme docs, API definition doc and the yaml files is all we need!

lewisdaly commented:
Great! Thanks @elnyry-sam-k .

lewisdaly commented:
Ok! This is now ready for review.

millerabel reviewed (commented):
Looks good to me, I'll defer to Sam and Michael for approval.

ehenrka commented:
I'm currently working on the review, but this will take some time. One general comment for now though is that could you please have a look at the formatting style used in FSP Interoperality API? As an example, for the Thirdparty API `POST /quotes` is used, while in FSP Interoperability **POST /quotes** is used. The full conventions can be found below or at https://docs.mojaloop.io/mojaloop-specification/fspiop-api/documents/API%20Definition%20v1.1.html#11-conventions-used-in-this-document:

|Type of Information|Convention|Example|
|---|---|---|
|**Elements of the API, such as resources**|Boldface|**/authorization**|
|**Variables**|Italics with in angle brackets|_{ID}_|
|**Glossary terms**|Italics on first occurrence; defined in _Glossary_|The purpose of the API is to enable interoperable financial transactions between a _Payer_ (a payer of electronic funds in a payment transaction) located in one _FSP_ (an entity that provides a digital financial service to an end user) and a _Payee_ (a recipient of electronic funds in a payment transaction) located in another FSP.|
|**Library documents**|Italics|User information should, in general, not be used by API deployments; the security measures detailed in _API Signature and API Encryption_ should be used instead.|

ehenrka commented:
Similarly for the styling of the sequence diagrams, could we use the same as in the FSP Interoperability API? The formatting was decided in the earlier CCB in #10.

ehenrka commented:
Another general comment is that it would be good if the documents had a similar basic structure. For example, the data-models-md document has a revision history, this is not present in the two patterns documents. The two patterns documents have a Contents section, this is not present in the data-models-md document.

I suggest you to have a look how all the documents under FSPIOP API are structured up to (including) Section 2. It would be great if we could have something similar here.

ehenrka commented:
Also recommend to have a look at the flows to tidy them up a bit before actual release, for example I found some " // This is new!" comments in some flow which is a bit confusing to a reader.

ehenrka commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+
+| Name | Cardinality | Type | Description |
+| --- | --- | --- | --- |
+| clientDataJSON    | 1 | string | JSON string with client data |
```
> | clientDataJSON    | 1 | string | JSON string with client data |

Please use a max length of the string (https://apisecurity.io/encyclopedia/content/oasv2/datavalidation/parameters/parameter-string-maxlength)

ehenrka commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+| Name | Cardinality | Type | Description |
+| --- | --- | --- | --- |
+| clientDataJSON    | 1 | string | JSON string with client data |
+| attestationObject | 1 | string | CBOR.encoded attestation object|
```
> | attestationObject | 1 | string | CBOR.encoded attestation object|

- Please use a max length of the string (https://apisecurity.io/encyclopedia/content/oasv2/datavalidation/parameters/parameter-string-maxlength)
- Please include a reference to what a CBOR.encoded object means

ehenrka commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+
+| Name | Cardinality | Type | Description |
+| --- | --- | --- | --- |
+| authenticatorData | 1 | string | Information about the authenticator. |
```
> | authenticatorData | 1 | string | Information about the authenticator. |

Please use a max length of the string (https://apisecurity.io/encyclopedia/content/oasv2/datavalidation/parameters/parameter-string-maxlength)

ehenrka commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+| Name | Cardinality | Type | Description |
+| --- | --- | --- | --- |
+| authenticatorData | 1 | string | Information about the authenticator. |
+| clientDataJSON    | 1 | string | base64 encoded JSON string containing information about the client. |
```
> | clientDataJSON    | 1 | string | base64 encoded JSON string containing information about the client. |

Please use a max length of the string (https://apisecurity.io/encyclopedia/content/oasv2/datavalidation/parameters/parameter-string-maxlength)

ehenrka commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+| --- | --- | --- | --- |
+| authenticatorData | 1 | string | Information about the authenticator. |
+| clientDataJSON    | 1 | string | base64 encoded JSON string containing information about the client. |
+| signature         | 1 | string | The signature generated by the private key associated with this credential. |
```
>| signature         | 1 | string | The signature generated by the private key associated with this credential. |

Please use a max length of the string (https://apisecurity.io/encyclopedia/content/oasv2/datavalidation/parameters/parameter-string-maxlength)

ehenrka commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+| authenticatorData | 1 | string | Information about the authenticator. |
+| clientDataJSON    | 1 | string | base64 encoded JSON string containing information about the client. |
+| signature         | 1 | string | The signature generated by the private key associated with this credential. |
+| userHandle        | 0..1 | string | This field is optionally provided by the authenticator, and represents the user.id that was supplied during registration.|
```
> | userHandle        | 0..1 | string | This field is optionally provided by the authenticator, and represents the user.id that was supplied during registration.|

- Please use a max length of the string (https://apisecurity.io/encyclopedia/content/oasv2/datavalidation/parameters/parameter-string-maxlength)
- "user.id" have never been used before as a term

ehenrka commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+| VERIFIED | The challenge was correctly signed. |
+| REJECTED | The challenge was not correctly signed. |
+##### 3.2.2.2 AuthorizationResponse
+The AuthorizationResponseType enumeration is the same as the AuthorizationResponse enumeration described in [Section 7.5.3](https://github.com/mojaloop/mojaloop-specification/blob/master/fspiop-api/documents/v1.1-document-set/API%20Definition%20v1.1.md#753-authorizationresponse) of Ref. 1 above.
```
> The AuthorizationResponseType enumeration ...

The name of the section is "AuthorizationResponse" (without "Type")

ehenrka commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+The ServiceType element contains a type of service where the requester wants a list of the participants in the scheme which provide that service. It must be a member of the appropriate enumeration.
+| Name | Cardinality | Type | Description |
+| --- | --- | --- | --- |
+| serviceType | 1 | Enum of String(1..32) | See [Section 3.2.2.10](#3229-servicetype) below ServiceType) for more information on allowed values. |
```
> | serviceType | 1 | Enum of String(1..32) | See [Section 3.2.2.10](#3229-servicetype) below ServiceType) for more information on allowed values. |

There is a ")" after ServiceType without a matching "("

ehenrka commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+| FIDO | The credential is based on a FIDO challenge. Its payload is a FIDOPublicKeyCredentialAttestation object. |
+| GENERIC | The credential is based on a simple public key validation. Its payload is a GenericCredential object |
+
+##### 3.2.2.7 PartyIdType
```
Suggesting to raise a change request on FSP Interoperability API to align the PartyIdType. At least I don't have anything against adding THIRD_PARTY_LINK.

ehenrka commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+
+| Name | Description |
+| ---  | ----------- |
+| accounts.getBalance | PISP can request a balance for the linked account|
```
> | accounts.getBalance | ...

- Please use UPPERCASE for enum values to be consistent with other enumerations. Suggesting ACCOUNTS_GET_BALANCE or similar
- I thought only transfer was supported so far?

ehenrka commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+| Name | Description |
+| ---  | ----------- |
+| accounts.getBalance | PISP can request a balance for the linked account|
+| accounts.transfer   | PISP can request a transfer of funds from the linked account in the DFSP|
```
> | accounts.transfer   | ...

Please use UPPERCASE for enum values to be consistent with other enumerations. Suggesting ACCOUNTS_TRANSFER or similar

ehenrka commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+| ---  | ----------- |
+| accounts.getBalance | PISP can request a balance for the linked account|
+| accounts.transfer   | PISP can request a transfer of funds from the linked account in the DFSP|
+| accounts.statement  | PISP can request a statement of individual transactions on a user’s account|
```
> | accounts.statement  | ...

- Please use UPPERCASE for enum values to be consistent with other enumerations. Suggesting ACCOUNTS_STATEMENT or similar
- I thought only transfer was supported so far?

lewisdaly replied:
We wanted to define the other future permissions that a DFSP can grant ahead of time to ensure good backwards compatibility.

ehenrka commented on thirdparty-api/transaction-patterns-transfer.md:
```diff
@@ -0,0 +1,327 @@
+* [1. Transfers](#Transfers)
+	* [1.1 Discovery](#Discovery)
+	* [1.2 Agreement](#Agreement)
+		* [1.2.1 `POST /thirdpartyRequests/transactions`](#POSTthirdpartyRequeststransactions)
```
> [1.2.1 `POST /thirdpartyRequests/transactions`](#POSTthirdpartyRequeststransactions)

Could we use the logical name for this section as well, as all other sections seems to be logical names?

ehenrka commented on thirdparty-api/transaction-patterns-transfer.md:
```diff
@@ -0,0 +1,327 @@
+    "responseType": "ACCEPTED",
+    "signedPayload": {
+        "signedPayloadType": "GENERIC",
+        "value": "utf-8 base64 encoded signature"
```
> "value": "utf-8 base64 encoded signature"

This seems to be "signedPayload" instead of value in the data model

ehenrka commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+| signedPayload | 0..1 | SignedPayload | If the `responseType` is `ACCEPTED`, `signedPayload` is required. |
+
+
+`SignedPayload` is defined as follows:
```
> `SignedPayload` is defined as follows:
This should be defined under Section 3.2, not as part of the definition of the callback.

ehenrka reviewed (commented):
I have mostly looked at the markdown files so far. Please have a look at the comments, and my more general comments as well.

lewisdaly commented:
Wow, thanks for this @ehenrka, I really appreciate your detailed review. Let me update the doc and get back to you - there will likely be changes in the swagger as well as a consequence of some of the changes you have suggested as well

ehenrka commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+The AuthenticationValue data element contains a response returned by the recipient of an authorization request. It is described in [Section 7.3.3](https://github.com/mojaloop/mojaloop-specification/blob/master/fspiop-api/documents/v1.1-document-set/API%20Definition%20v1.1.md#733-authenticationvalue) of Ref. 1 above, and is extended to support the new authentication type used for PISP. The data mo ...
+| Name | Cardinality | Type | Description |
+| --- | --- | --- | --- |
+| AuthenticationValue | 1 | `OtpValue` or `String(1..64)` or `BinaryString` | Contains the authentication value. The format depends on the authentication type used in the AuthenticationInfo complex type. If OTP: OtpValue; If QRCODE: String(1..64); If U2F: BinaryString |
```
Very sorry, seems like I was blind and missed the BinaryString in the Type column

lewisdaly commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+## 3 Third Party API
+
+This section describes the content of the API which will be used by PISPs.
+The content of the API falls into two sections. The first section manages the process for linking customer accounts and providing a general permission mechanism for PISPs to perform operations on those accounts. The second section manages the transfer of funds at the instigation of a PISP.
```
Yeah these things were written at different times and not always updated nicely in step with one another. 

Following the pattern from the FSPIOP-API, I'm going to remove the specific talk of the usage of the API in favour of referencing the transaction-patterns document.

lewisdaly commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+Party’s accounts, defined by an identifier `<ID>`, where `<ID>` is some identifier a user
+uses to access their account with the DFSP, such as a phone number, email address, clientId.
+
+> __Note:__ This differs from the `GET /parties/<TYPE>/<ID>` request of Ref.1. above
```
Ok, I can remove it.

lewisdaly commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+
+The HTTP request `GET /accounts/<ID>` is used to lookup information about the requested
+Party’s accounts, defined by an identifier `<ID>`, where `<ID>` is some identifier a user
+uses to access their account with the DFSP, such as a phone number, email address, clientId.
```
I'll remove it, it's just a vague example.

lewisdaly commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+Used by: PISP
+
+The HTTP request `GET /accounts/<ID>` is used to lookup information about the requested
+Party’s accounts, defined by an identifier `<ID>`, where `<ID>` is some identifier a user
```
I'll replace it with "user", since that is more specific in this case.

lewisdaly commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+
+This section describes the services that a PISP can request on the /accounts resource.
+
+###### 3.1.1.1.1  `GET /accounts/<ID>`
```
No. The idea here is that we are using some opaque identifier that is known to the DFSP, and we don't concern ourselves with Type of the identifier. Since the user already has selected their DFSP, we don't need to do a Party Lookup to determine the DFSP the request should be routed to, we don't need to conform to the FSPIOP /Type/Id(/subId) format. 

Really this allows us to ask the question: "What do you use to log in with your bank?"

From a few lines above:
> where `<ID>` is some identifier a user uses to access their account with the DFSP, such as a phone number, email address, clientId.

lewisdaly commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+
+The responses for the `/accounts` resource are as follows
+
+###### 3.1.1.2.1  `PUT /accounts/<ID>`
```
Discussed above.

lewisdaly commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+| --- | --- | --- | --- |
+| accountList | 1 | AccountList | Information about the accounts that the DFSP associates with the identifier sent by the PISP. |
+
+###### 3.1.1.2.2  `PUT /accounts/<ID>/error`
```
Discussed above

lewisdaly commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+
+| Name | Cardinality | Type | Description |
+| --- | --- | --- | --- |
+| accountList | 1 | AccountList | Information about the accounts that the DFSP associates with the identifier sent by the PISP. |
```
It would be, but as soon as something changes in this doc the link will break. I noticed that the other data-models doc doesn't have such a link.

lewisdaly commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+Used by: PISP
+
+The HTTP request `GET /consentRequests/<ID>` is used to get information about a previously 
+requested consent. The `<ID>` in the URI should contain the requestId that was assigned to the 
```
Yep, it's meant to be `consentRequestId`

lewisdaly commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+| userId           | 1 | string        | The identifier used in the `GET /accounts/<ID>`. Used by the DFSP to correlate an account lookup to a `consentRequest` |
+| scopes | 1..n | Scope | One or more requests for access to a particular account. In each case, the address of the account and the types of access required are given. |
+| authChannels | 1..n | ConsentRequestChannelType | A collection of the types of authentication that the DFSP may use to verify that its customer has in fact requested access for the PISP to the accounts requested. |
+| callbackUri | 1 | Uri |The callback URI that the user will be redirected to after completing verification via the WEB authorization channel |
```
Yes. The PISP must provide a callbackUri because it doesn't know ahead of time which AuthChannel the DFSP will choose to use. I'll add a note to that effect

lewisdaly commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+| Name | Cardinality | Type | Description |
+| --- | --- | --- | --- |
+| scopes       | 1..n | Scope | One or more requests for access to a particular account. In each case, the address of the account and the types of access required are given. |
+| authChannels | 1 | ConsentRequestChannelType | A list of one element, which the DFSP uses to inform |
```
oops.

lewisdaly commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+| consentId | 1 | CorrelationId | Common ID between the PISP and the Payer DFSP for the consent object. The ID should be reused for resends of the same consent. A new ID should be generated for each new consent.|
+| consentRequestId | 1 | CorrelationId | The ID given to the original consent request on which this consent is based. |
+| scopes   | 1..n  | Scope |One or more accounts on which the DFSP is prepared to grant specified permissions to the PISP. |
+| credential | 0..1 | Credential | The credential which is being used to support the consents. |
```
It wouldn't but the `Consent` is issued by the DFSP, and the credential is attached to the consent by the PISP. 

By allowing the DFSP to specify the credential ahead of time (in the `POST /consents`) we allow the DFSP to send the whole Consent + credential object to the Auth Service for later validation, and also will be able to support future use cases where a DFSP collects the credential on behalf of the PISP.

lewisdaly commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+The `PUT /parties` resource will use the same form as the resource described in 
+[Section 6.3.4.1](https://github.com/mojaloop/mojaloop-specification/blob/master/fspiop-api/documents/v1.1-document-set/API%20Definition%20v1.1.md#6341-put-partiestypeid) of Ref. 1 above.
+
+It should be noted, however, that the Party object returned from this resource has a different
```
Yep, I think I missed updating this in my last pass. We don't need to change the object any more, I'll update this text.

lewisdaly commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+
+| Name | Cardinality | Type | Description |
+| --- | --- | --- | --- |
+| providers | 1…n | FspId |A list of the Ids of the participants who provide the service requested. |
```
Sure, thanks for these suggestions.

lewisdaly commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+The AuthenticationInfo data type used in these definitions is as defined in [Section 7.4.1](https://github.com/mojaloop/mojaloop-specification/blob/master/fspiop-api/documents/v1.1-document-set/API%20Definition%20v1.1.md#741-authenticationinfo) of Ref. 1 above.
+##### 3.2.1.5 AuthenticationResponse
+##### 3.2.1.6 AuthenticationValue
+The AuthenticationValue data element contains a response returned by the recipient of an authorization request. It is described in [Section 7.3.3](https://github.com/mojaloop/mojaloop-specification/blob/master/fspiop-api/documents/v1.1-document-set/API%20Definition%20v1.1.md#733-authenticationvalue) of Ref. 1 above, and is extended to support the new authentication type used for PISP. The data mo ...
```
We actually don't need this anymore.

lewisdaly commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+##### 3.2.1.21 PartyIdInfo
+The PartyIdInfo data type used in these definitions is as defined in [Section 7.4.13](https://github.com/mojaloop/mojaloop-specification/blob/master/fspiop-api/documents/v1.1-document-set/API%20Definition%20v1.1.md#7413-partyidinfo) of Ref. 1 above.
+
+##### 3.2.1.22 Quote 
```
Yep!

lewisdaly commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+> from a list of supported DFSPs, so the switch does not need to perform a lookup to determine
+> the DFSP this request should be routed to.  
+
+See Section 5.1.6.11 of Ref. 1 above for more information regarding addressing of a Party.
```
I've removed this reference as it's outdated.

lewisdaly commented on thirdparty-api/data-models.md:
```diff
@@ -0,0 +1,953 @@
+#### 3.1.1 `/accounts`
+
+The `/accounts` resource is used to request information from a DFSP relating to the accounts 
+it holds for a given identifier. The identifier is given to the PISP by the user, and the 
```
We do below, in section 3.1.1.1.1

lewisdaly commented on thirdparty-api/transaction-patterns-linking.md:
```diff
@@ -0,0 +1,403 @@
+
+In this phase, a PISP Server needs to know what DFSPs are available to link
+with. This is *unlikely* to be done on-demand (e.g., when a User clicks "link"
+in the PISP mobile App), and far more likely to be done periodically and cached
```
Since this is an Async API, I guess we could also consider a case where the Switch sends a `PUT /services/THIRD_PARTY_DFSP` call to all PISPs whenever a new participant is added.

lewisdaly commented on thirdparty-api/transaction-patterns-linking.md:
```diff
@@ -0,0 +1,403 @@
+
+In the OTP authentication channel, the DFSP sends an 'out of bound' OTP message
+to their user (e.g. over SMS or Email). The PISP prompts the user for this OTP
+message, and includes it in the `authToken` field in the `PUT /consentRequests/{id}`
```
Yep, you're right.

lewisdaly commented on thirdparty-api/transaction-patterns-linking.md:
```diff
@@ -0,0 +1,403 @@
+
+![Authentication (Web)](./assets/linking/3-authentication-web.svg)
+
+#### DFSPAuthorizeSimulator
```
I agree. Removed.

lewisdaly commented on thirdparty-api/transaction-patterns-linking.md:
```diff
@@ -0,0 +1,403 @@
+#### DFSPAuthorizeSimulator
+DFSPAuthorizeSimulator is api service for demo purposes. For the `Web flow` it is the backend  for single page application (SPA) which demonstrates the process of User's authorizations and granting the consents at DFSP web service.
+
+It is implemented as extension for TestingToolkit service which allows to define testing scenarios as TTK rules driven by `username` parameter.
```
I'm just going to remove this section

lewisdaly commented on thirdparty-api/transaction-patterns-linking.md:
```diff
@@ -0,0 +1,403 @@
+
+![Authentication (OTP)](./assets/linking/3-authentication-otp.svg)
+
+#### DFSPAuthorizeSimulator
```
yep, I agree. Removed.

lewisdaly commented on thirdparty-api/transaction-patterns-linking.md:
```diff
@@ -0,0 +1,403 @@
+for how the User can prove that it consents to each individual transfer in the
+future.
+
+This phase consists exclusively of the DFSP requesting that a new consent be
```
Yeah I'm not sure. I'll tidy this bit up.

lewisdaly commented on thirdparty-api/transaction-patterns-linking.md:
```diff
@@ -0,0 +1,403 @@
+
+## 3.2. Credential registration
+
+TODO!
```
Removed this section

lewisdaly commented on thirdparty-api/transaction-patterns-transfer.md:
```diff
@@ -0,0 +1,327 @@
+
+#### <a name='ThirdpartyAuthorizationRequest'></a>1.2.2 Thirdparty Authorization Request
+
+The DFSP will then issue a quotation request (`POST /quotes`) to the Payee DFSP. Upon receiving the `PUT /quotes/{Id}` callback from the Payee DFSP, the Payer DFSP needs to confirm the details of the transaction with the PISP.
```
Would saying something like "The DFSP may then issue..." suffice?

lewisdaly commented:
@ehenrka it's ready to go again for another review. 

I think I addressed most of your comments, and I marked those I was sure I addressed as "resolved".

I still am yet to update the swagger files, but I'll wait until you are happy with the `.md` and `puml` docs

lewisdaly commented:
Thanks @ehenrka. I found I can't comment on some of the comments you made, so I'll include some replies below:


> Thanks! Can we describe this somewhere in the document so that it more clear?

Yep, Credential is defined in section 3.2.1.10


> Should be fairly simple to just do a search and replace of instance of "/{ID}" to "/_{ID}_"..
I think I've got them all this time.

> No, it should preferably not send a PUT as it is used for callbacks today. Something along the following is how I would suggest it to be:...

I'd rather we just suggest to the PISP to call the `GET /services` resource once a day and cache the results. That seems much simpler than defining and implementing such a subscription mechanism.

elnyry-sam-k commented:
@lewisdaly - as we discussed in the CCB meeting, is this now ready to go? (can you please update the branch etc).. @ehenrka any further comments?

ehenrka commented:
I'm happy with the updates from @lewisdaly.

lewisdaly commented:
Great! Thanks @ehenrka. @elnyry-sam-k I guess we are good to merge in then...

elnyry-sam-k commented:
Thanks @lewisdaly , please go ahead!
